testing.md

```diff
@@ -1 +1,153 @@
 # Testing
+
+## Testing strutturale (white-box)
+
+Il **testing strutturale** tiene conto della struttura interna del programma con l'obiettivo di sollecitare tutte le parti del programma.
+
+### Criterio di copertura delle istruzioni (statement coverage)
+
+Si vuole fare in modo che ogni istruzione del programma venga eseguita almeno una volta.
+Consideriamo la seguente funzione:
+
+```java
+int sqrt(int n) {
+  int i = 1, z = 1;
+  if(n < 0) 
+    return -1;
+  while(z <= n) {
+    i = i + 1;
+    z = i * i;
+  }
+  return i - 1;
+}
+```
+
+Dobbiamo selezionare dei casi di testi in modo che ogni istruzione venga eseguita almeno una volta.
+possiamo scegliere i seguenti casi di test:
+
+- `n = -1` per eseguire il corpo dell'`if`
+- `n = 1` per eseguire il corpo del `while`
+
+_Osservazione_: se una funzione ha n `return` allora avremo per forza **almeno n casi di test**.
+
+### Criterio di copertura delle decisioni (edge coverage)
+
+Si vuole coprire tutte le possibili _diramazioni_ del programma. per fare ciò è utile costruire il **control flow graph** (o **CFG**) del programma in cui si evidenziono proprio tutte le diramazioni.
+Nella figura seguente è riportato il CFG della funzione `sqrt` (esempio precedente):
+
+![](./immagini/cfg.svg)
+
+Ogni arco rappresenta una possibile diramazione. Si noti che ogni `if` e `while` sono rappresentati da due archi, uno per il caso `true` e uno per il caso `false` (anche se ad esempio il ramo `else` non è esplicitamente presente nel codice).
+
+Il criterio **edge coverage** vuole che ogni arco del CFG venga eseguito almeno una volta. Per fare ciò possiamo scegliere i seguenti casi di test:
+
+- `n = -1` per eseguire solo l'arco `n < 0` (e finire il programma)
+- `n = 1` per eseguire il ramo else del primo `if` e per entrare nel `while` (e quindi eseguire tutti e due gli archi del `while`)
+
+### Criterio di copertura delle decisioni e delle condizioni (edge and condition coverage)
+
+L’insieme dei casi di test deve essere definito in modo che ogni ramo del CFG venga attraversato almeno una volta e con **tutti i possibili valori** nelle sottoespressioni che compaiono nelle condizioni composte.
```

Comment:
Cambierei:
con **tutti i possibili valori** nelle sottoespressioni che compaiono nelle condizioni composte -> sollecitando tutti i valori di verità delle sottoespressioni che compaiono nelle condizioni composte, tenendo conto della _short circuit evaluation_

Perché ad esempio, nello snippet di codice sotto "**tutti i possibili valori** nelle sottoespressioni" sono 4 (la prima vera e la seconda falsa, il viceversa, entrambe vere, entrambe false) ma ne bastano 3 per sollecitare tutti i valori di verità

Poi più in là pensavo di rimetterci mano a questa parte proprio per rendere più chiara questa cosa, perché anche come l'ha spiegata San Pietro non mi convince, però per ora va benissimo così, poi se ho tempo me ne occuperò io alla fine